File: app-ios/TutanotaSharedFramework/Notifications/EncryptedRepeatRule.swift
import Foundation
import tutasdk

public enum RepeatPeriod: Int, SimpleStringDecodable {
	case daily = 0
	case weekly
	case monthly
	case annually
}

public enum EndType: Int, SimpleStringDecodable {
	case never = 0
	case count
	case untilDate
}

public enum RepeatEndCondition: Equatable {
	case never
	case count(times: Int64)
	case untilDate(date: Date)

	public init(endType: EndType, endValue: Int64) {
		switch endType {
		case .never: self = .never
		case .count: self = .count(times: endValue)
		case .untilDate:
			let seconds = Double(endValue) / 1000
			self = .untilDate(date: Date(timeIntervalSince1970: seconds))
		}
	}
}

public enum ByRuleType: String, Codable, Equatable, SimpleStringDecodable {
	case byminute
	case byhour
	case byday
	case bymonthday
	case byyearday
	case byweekno
	case bymonth
	case bysetpos
	case wkst

	public init(value: String) throws {
		switch value {
		case "0": self = .byminute
		case "1": self = .byhour
		case "2": self = .byday
		case "3": self = .bymonthday
		case "4": self = .byyearday
		case "5": self = .byweekno
		case "6": self = .bymonth
		case "7": self = .bysetpos
		case "8": self = .wkst
		default: throw TUTErrorFactory.createError("Invalid ByRuleType")
		}
	}
}

public struct EncryptedDateWrapper: Codable, Hashable {
    public let date: Base64

    enum CodingKeys: String, CodingKey {
        case date = "2075"
    }
}
public struct EncryptedAdvancedRuleWrapper: Codable, Hashable {
	public let ruleType: String
	public let interval: String

	enum CodingKeys: String, CodingKey {
        case interval = "2524"
        case ruleType = "2523"
    }
}
public struct AdvancedRule: Codable, Hashable {
	public let ruleType: ByRuleType
	public let interval: String

	public func toSDKRule() -> ByRule { ByRule(byRule: self.ruleType.toSDKType(), interval: self.interval) }
}

extension ByRuleType {
	func toSDKType() -> tutasdk.ByRuleType {
		switch self {
		case .byminute: return tutasdk.ByRuleType.byMinute
		case .byhour: return tutasdk.ByRuleType.byHour
		case .byday: return tutasdk.ByRuleType.byDay
		case .bymonth: return tutasdk.ByRuleType.byMonth
		case .bymonthday: return tutasdk.ByRuleType.byMonthday
		case .byyearday: return tutasdk.ByRuleType.byYearDay
		case .byweekno: return tutasdk.ByRuleType.byWeekNo
		case .bysetpos: return tutasdk.ByRuleType.bySetPos
		case .wkst: return tutasdk.ByRuleType.wkst
		}
	}
}

public struct EncryptedRepeatRule: Codable, Hashable {
	public let frequency: Base64
	public let interval: Base64
	public let timeZone: Base64
	public let endType: Base64
	public let endValue: Base64?
	public let excludedDates: [EncryptedDateWrapper]
	public let advancedRules: [EncryptedAdvancedRuleWrapper]

	public init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)
		if let excludedDates = try container.decodeIfPresent([EncryptedDateWrapper].self, forKey: .excludedDates) {
			self.excludedDates = excludedDates
		} else {
			self.excludedDates = [EncryptedDateWrapper]()
		}

		if let advancedRules = try container.decodeIfPresent([EncryptedAdvancedRuleWrapper].self, forKey: .advancedRules) {
			self.advancedRules = advancedRules
		} else {
			self.advancedRules = [EncryptedAdvancedRuleWrapper]()
		}

		self.frequency = try container.decode(Base64.self, forKey: .frequency)
		self.interval = try container.decode(Base64.self, forKey: .interval)
		self.timeZone = try container.decode(Base64.self, forKey: .timeZone)
		self.endType = try container.decode(Base64.self, forKey: .endType)

		if let endValue = try container.decodeIfPresent((Base64?).self, forKey: .endValue) { self.endValue = endValue } else { self.endValue = nil }
	}


    enum CodingKeys: String, CodingKey {
        case frequency = "1559"
        case interval = "1562"
        case timeZone = "1563"
        case endType = "1560"
        case endValue = "1561"
        case excludedDates = "2076"
        case advancedRules = "2525"
    }
}

File: app-android/tutashared/src/main/java/de/tutao/tutashared/alarms/EncryptedRepeatRule.kt
package de.tutao.tutashared.alarms

import androidx.room.TypeConverter
import androidx.room.TypeConverters
import de.tutao.tutasdk.ByRule
import de.tutao.tutasdk.ByRuleType
import de.tutao.tutashared.AndroidNativeCryptoFacade
import de.tutao.tutashared.decryptDate
import de.tutao.tutashared.decryptNumber
import de.tutao.tutashared.decryptString
import kotlinx.serialization.Serializable
import java.util.TimeZone

@Serializable
class EncryptedDateWrapper(
	val date: String
)

@Serializable
class EncryptedByRuleWrapper(
	val interval: String,
	val ruleType: String
)

@Serializable
@TypeConverters(
	EncryptedRepeatRule.ExcludedDateWrapperConverter::class,
	EncryptedRepeatRule.ByRuleWrapperConverter::class
)
class EncryptedRepeatRule(
	val frequency: String,
	val interval: String,
	val timeZone: String,
	val endType: String,
	val endValue: String?,
	val excludedDates: List<EncryptedDateWrapper>,
	val advancedRules: List<EncryptedByRuleWrapper>
) {
	internal class ExcludedDateWrapperConverter {
		@TypeConverter
		fun listToString(excludedDatesList: List<EncryptedDateWrapper>) =
			excludedDatesList.joinToString(",") { it.date }

		@TypeConverter
		fun stringToList(string: String?) = if (string != null && string.isNotEmpty()) string.split(",")
			.map { EncryptedDateWrapper(it) } else emptyList()
	}

	internal class ByRuleWrapperConverter {
		@TypeConverter
		fun listToString(rules: List<EncryptedByRuleWrapper>) =
			rules.joinToString(";") { "${it.interval},${it.ruleType}" }

		@TypeConverter
		fun stringToList(string: String?) = if (!string.isNullOrEmpty()) {
			string.split(";").map {
				val (interval, ruleType) = it.split(",")
				EncryptedByRuleWrapper(interval, ruleType)
			}
		} else emptyList<EncryptedByRuleWrapper>()
	}
}

fun EncryptedRepeatRule.decrypt(crypto: AndroidNativeCryptoFacade, sessionKey: ByteArray): RepeatRule {
	val repeatPeriodNumber = crypto.decryptNumber(frequency, sessionKey)
	val repeatPeriod = RepeatPeriod[repeatPeriodNumber]

	val endTypeNumber = crypto.decryptNumber(endType, sessionKey)
	val endType = EndType[endTypeNumber]
	return RepeatRule(
		frequency = repeatPeriod,
		interval = crypto.decryptNumber(interval, sessionKey).toInt(),
		timeZone = TimeZone.getTimeZone(crypto.decryptString(timeZone, sessionKey)),
		endValue = if (endValue != null) crypto.decryptNumber(endValue, sessionKey) else null,
		endType = endType,
		excludedDates = excludedDates.map { crypto.decryptDate(it.date, sessionKey) },
		advancedRules = advancedRules.map {
			val interval = crypto.decryptString(it.interval, sessionKey)
			val rule = ByRuleType.fromValue(crypto.decryptNumber(it.ruleType, sessionKey).toInt())

			if (rule == ByRuleType.BY_SET_POS) {
				assertNumericSetPosInterval(interval)
			}

			ByRule(rule, interval)
		}
	)
}

private fun assertNumericSetPosInterval(interval: String) {
	try {
		interval.toInt()
	} catch (e: NumberFormatException) {
		throw NumberFormatException("Invalid bySetPos rule with interval $interval")
	}
}

fun ByRuleType.Companion.fromValue(value: Int) = run {
	when (value) {
		0 -> ByRuleType.BY_MINUTE
		1 -> ByRuleType.BY_HOUR
		2 -> ByRuleType.BY_DAY
		3 -> ByRuleType.BY_MONTHDAY
		4 -> ByRuleType.BY_YEAR_DAY
		5 -> ByRuleType.BY_WEEK_NO
		6 -> ByRuleType.BY_MONTH
		7 -> ByRuleType.BY_SET_POS
		8 -> ByRuleType.WKST

		else -> {
			throw Exception("Invalid rule type")
		}
	}
}